Add SupRemo in Remote_Mgmt

dxdemetriou opened the pull request:
Ref: https://github.com/glpi-project/glpi-agent/issues/35

dxdemetriou committed:
Add SupRemo in Remote_Mgmt

Ref: https://github.com/glpi-project/glpi-agent/issues/35

dxdemetriou committed:
Update SupRemo.pm

g-bougard commented on lib/GLPI/Agent/Task/Inventory/Generic/Remote_Mgmt/SupRemo.pm:
```diff
@@ -0,0 +1,129 @@
+        # Depending on the installation the supremo key can be in two place in X64 OS
+
+        my $key = getRegistryKey(
+            path => "HKEY_LOCAL_MACHINE/SOFTWARE/Wow6432Node/Supremo",
```
It seems you're requesting the same key than on line 30. Do you forgot to remove one `Wow6432Node` from the key path here ?
```suggestion
            path => "HKEY_LOCAL_MACHINE/SOFTWARE/Supremo",
```

g-bougard replied:
This is still wrong as on 64bit OS, if the value doesn't exist, you'll retry the same key at l.30.
I don't know Supremo, but can I guess they only provide at the moment a 32bit version ? If yes, just suppose they will provide a 64bit version in the future. And then you need to fix this code.

g-bougard commented on lib/GLPI/Agent/Task/Inventory/Generic/Remote_Mgmt/SupRemo.pm:
```diff
@@ -0,0 +1,129 @@
+    GLPI::Agent::Tools::Win32->use();
+
+    my $clientid = getRegistryValue(
+        path => "HKEY_LOCAL_MACHINE/SOFTWARE/Supremo/Wow6432Node/ClientID",
```
As above.
```suggestion
        path => "HKEY_LOCAL_MACHINE/SOFTWARE/Supremo/ClientID",
```

g-bougard replied:
Same remark as above.

g-bougard commented on lib/GLPI/Agent/Task/Inventory/Generic/Remote_Mgmt/SupRemo.pm:
```diff
@@ -0,0 +1,129 @@
+
+    unless ($clientid) {
+        my $supremover_reg = getRegistryKey(
+            path => "HKEY_LOCAL_MACHINE/SOFTWARE/Wow6432Node/Supremo",
```
As above
```suggestion
            path => "HKEY_LOCAL_MACHINE/SOFTWARE/Supremo",
```

dxdemetriou replied:
It's needed both on 32bit and 64bit the Wow6432Node. I just left both architectures on the code in case of future updates.

g-bougard replied:
If you install 32bit version on 32bit OS or 64bit version on 64bit OS, the keypath should probably be `HKEY_LOCAL_MACHINE/SOFTWARE/Supremo`, and if you install 32bit version on 64bit OS, the keypath would be `HKEY_LOCAL_MACHINE/SOFTWARE/Wow6432Node/Supremo`. `Wow6432Node` node doesn't exist on 32bit OS.

dxdemetriou replied:
Yes, sorry, for some reason I thought the opposite :confused: . I have fixed them, thanks.

g-bougard commented on lib/GLPI/Agent/Task/Inventory/Generic/Remote_Mgmt/SupRemo.pm:
```diff
@@ -0,0 +1,129 @@
+    my $supRemoID = _getID(
+        osname  => OSNAME,
+        logger  => $logger
+    );
```
Don't pass osname here but directly use OSNAME to call right sub
```suggestion
    my $supRemoID = OSNAME eq 'MSWin32' ?
        _getID_MSWin32(logger  => $logger) : _getID_supremo_info(logger  => $logger);
```

g-bougard commented on lib/GLPI/Agent/Task/Inventory/Generic/Remote_Mgmt/SupRemo.pm:
```diff
@@ -0,0 +1,129 @@
+    return _getID_MSWin32() if $osname eq "MSWin32";
+
+    return _getID_supremo_info(%params);
+}
```
Remove this with the simpler code I propose to call the right sub
```suggestion
```

g-bougard commented on lib/GLPI/Agent/Task/Inventory/Generic/Remote_Mgmt/SupRemo.pm:
```diff
@@ -0,0 +1,129 @@
+            path => "HKEY_LOCAL_MACHINE/SOFTWARE/Wow6432Node/Supremo",
+            # Important for remote inventory optimization
+            required    => [ qw/ClientID/ ],
+        );
```
You should add maxdepth here
```suggestion
            required    => [ qw/ClientID/ ],
            maxdepth    => 2,
        );
```

g-bougard commented on lib/GLPI/Agent/Task/Inventory/Generic/Remote_Mgmt/SupRemo.pm:
```diff
@@ -0,0 +1,129 @@
+                path => "HKEY_LOCAL_MACHINE/SOFTWARE/Wow6432Node/Supremo",
+                # Important for remote inventory optimization
+                required    => [ qw/ClientID/ ],
+            );
```
You should add maxdepth here
```suggestion
                required    => [ qw/ClientID/ ],
                maxdepth    => 2,
            );
```

g-bougard commented on lib/GLPI/Agent/Task/Inventory/Generic/Remote_Mgmt/SupRemo.pm:
```diff
@@ -0,0 +1,129 @@
+            last if (defined($clientid));
+        }
+    }
+
```
Here you must handle correctly the case clientid is not found to avoid an error on sprintf() call and just ignore the inventory.
```suggestion

    return unless $clientid;

```

g-bougard commented on lib/GLPI/Agent/Task/Inventory/Generic/Remote_Mgmt/SupRemo.pm:
```diff
@@ -0,0 +1,129 @@
+    my (%params) = @_;
+
+    return getFirstMatch(
+        command => "supremo --info",
+        pattern => qr/SupRemo ID:(?:\033\[0m|\s)*(\d+)\s+/,
```
To avoid special chars in the regexp, can you check if there's an option for `supremo` command to have a machine readable output or just escape the pretty formatting ?

dxdemetriou replied:
There is no option from the executable to have a readable output

g-bougard reviewed (changes requested):
Hi @dxdemetriou 

thank you for your submission.

I think you made few mistakes with the registry path. Can you check my comments and fix them ?

I also suggest you few other changes.

g-bougard commented on lib/GLPI/Agent/Task/Inventory/Generic/Remote_Mgmt/SupRemo.pm:
```diff
@@ -0,0 +1,129 @@
+sub _getID_supremo_info {
+    my (%params) = @_;
+
+    return getFirstMatch(
```
Another point here, just to be sure: Is the clientid formatted as expected here like a number on 9 digits prefixed by zeros as the returned value of _getID_MSWin32() is suggesting it must be ?

dxdemetriou replied:
Yes, if the prefixed zeroes are missing the sofware does not load the number and gives an error

dxdemetriou committed:
Update lib/GLPI/Agent/Task/Inventory/Generic/Remote_Mgmt/SupRemo.pm

Don't pass osname here but directly use OSNAME to call right sub

Co-authored-by: Guillaume Bougard <gbougard@teclib.com>

dxdemetriou committed:
Update lib/GLPI/Agent/Task/Inventory/Generic/Remote_Mgmt/SupRemo.pm

Remove this with the simpler code I propose to call the right sub

Co-authored-by: Guillaume Bougard <gbougard@teclib.com>

dxdemetriou committed:
Update lib/GLPI/Agent/Task/Inventory/Generic/Remote_Mgmt/SupRemo.pm

maxdepth

Co-authored-by: Guillaume Bougard <gbougard@teclib.com>

dxdemetriou committed:
Update lib/GLPI/Agent/Task/Inventory/Generic/Remote_Mgmt/SupRemo.pm

maxdepth

Co-authored-by: Guillaume Bougard <gbougard@teclib.com>

dxdemetriou committed:
Update lib/GLPI/Agent/Task/Inventory/Generic/Remote_Mgmt/SupRemo.pm

Here you must handle correctly the case clientid is not found to avoid an error on sprintf() call and just ignore the inventory.

Co-authored-by: Guillaume Bougard <gbougard@teclib.com>

dxdemetriou committed:
Update lib/GLPI/Agent/Task/Inventory/Generic/Remote_Mgmt/SupRemo.pm

Co-authored-by: Guillaume Bougard <gbougard@teclib.com>

dxdemetriou committed:
Update lib/GLPI/Agent/Task/Inventory/Generic/Remote_Mgmt/SupRemo.pm

Co-authored-by: Guillaume Bougard <gbougard@teclib.com>

dxdemetriou committed:
Update lib/GLPI/Agent/Task/Inventory/Generic/Remote_Mgmt/SupRemo.pm

Co-authored-by: Guillaume Bougard <gbougard@teclib.com>

g-bougard commented:
Thank you for your contribution ;-)
